provider/provider.go

```diff
@@ -124,57 +123,99 @@ func (p *dockerNativeProvider) StreamInvoke(
 // the provider inputs are using for detecting and rendering diffs.
 func (p *dockerNativeProvider) Check(ctx context.Context, req *rpc.CheckRequest) (*rpc.CheckResponse, error) {
 	urn := resource.URN(req.GetUrn())
-	label := fmt.Sprintf("%s.Create(%s)", p.name, urn)
+	label := fmt.Sprintf("%s.Check(%s)", p.name, urn)
 	logging.V(9).Infof("%s executing", label)
 
-	inputs, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{
-		KeepUnknowns: false,
+	news := req.GetNews()
+
+	inputs, err := plugin.UnmarshalProperties(news, plugin.MarshalOptions{
+		KeepUnknowns: true,
 		SkipNulls:    true,
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	// Set defaults
+	//if inputs["build"].ContainsUnknowns() {
+	//	//// We skip some of the "nice-to-have" default and verification logic in the case of unknowns.
+	//	//// This should be fine, given that _any_ unknowns in the Build field should trigger a diff.
+	//	//// Furthermore, all of this will get called again during `pulumi up`.
+	//	//inputStruct, err := plugin.MarshalProperties(inputs, plugin.MarshalOptions{
+	//	//	KeepUnknowns: true,
+	//	//	SkipNulls:    true,
+	//	//})
+	//	//if err != nil {
+	//	//	return nil, err
+	//	//}
+	//
+	//	return &rpc.CheckResponse{Inputs: news, Failures: nil}, nil
+	//}
+
+	buildOnPreview := marshalBuildOnPreview(inputs)
+	inputs["buildOnPreview"] = resource.NewBoolProperty(buildOnPreview)
+
 	build, err := marshalBuildAndApplyDefaults(inputs["build"])
 	if err != nil {
 		return nil, err
 	}
+	// Set the resource inputs to the default values
+	var knownDockerfile bool
+	if inputs["build"].IsNull() {
+		inputs["build"] = resource.NewObjectProperty(resource.PropertyMap{
+			"dockerfile": resource.NewStringProperty(build.Dockerfile),
+			"context":    resource.NewStringProperty(build.Context),
+		})
+	} else {
+		// We do not want to set these fields if their values are Unknown.
+		if !inputs["build"].ObjectValue()["dockerfile"].ContainsUnknowns() {
+			inputs["build"].ObjectValue()["dockerfile"] = resource.NewStringProperty(build.Dockerfile)
+			knownDockerfile = true
+		}
+		if !inputs["build"].ObjectValue()["context"].ContainsUnknowns() {
+			inputs["build"].ObjectValue()["context"] = resource.NewStringProperty(build.Context)
+		}
+	}
 
 	// Verify Dockerfile at given location
-	if _, statErr := os.Stat(build.Dockerfile); statErr != nil {
-		if filepath.IsAbs(build.Dockerfile) {
-			return nil, fmt.Errorf("could not open dockerfile at absolute path %s: %v", build.Dockerfile, statErr)
+	if knownDockerfile {
+
+		if _, statErr := os.Stat(build.Dockerfile); statErr != nil {
+			if filepath.IsAbs(build.Dockerfile) {
+				return nil, fmt.Errorf("could not open dockerfile at absolute path %s: %v", build.Dockerfile, statErr)
+			}
```

Comment:
If `inputs["build"]` is "null", then I think we use default values on lines 164-167. But once we get down here I don't think we'll check if the default `Dockerfile` actually exists because `knownDockerfile` is still `false` -- is that intentional?

Reply:
It is not - thank you for the catch!